python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -744,31 +751,83 @@ def _submit_runs_and_update_backfill_in_chunks(
     yield backfill_data_with_submitted_runs
 
 
-def execute_asset_backfill_iteration(
+def _check_target_partitions_subset_is_valid(
+    asset_key: AssetKey,
+    asset_graph: AssetGraph,
+    target_partitions_subset: Optional[PartitionsSubset],
+    instance_queryer: CachingInstanceQueryer,
+) -> None:
+    """Checks for any partitions definition changes since backfill launch that should mark
+    the backfill as failed.
+    """
+    if asset_key not in asset_graph.all_asset_keys:
+        raise DagsterDefinitionChangedDeserializationError(
+            f"Asset {asset_key} existed at storage-time, but no longer does"
+        )
+
+    partitions_def = asset_graph.get_partitions_def(asset_key)
+
+    if target_partitions_subset:  # Asset was partitioned at storage time
+        if partitions_def is None:
+            raise DagsterDefinitionChangedDeserializationError(
+                f"Asset {asset_key} had a PartitionsDefinition at storage-time, but no longer"
+                " does"
+            )
+
+        # If the asset was time-partitioned at storage time but the time partitions def
+        # has changed, mark the backfill as failed
+        if isinstance(
+            target_partitions_subset, TimeWindowPartitionsSubset
+        ) and target_partitions_subset.partitions_def.get_serializable_unique_identifier(
+            instance_queryer
+        ) != partitions_def.get_serializable_unique_identifier(instance_queryer):
+            raise DagsterDefinitionChangedDeserializationError(
+                f"This partitions definition for asset {asset_key} has changed since this backfill"
+                " was stored. Changing the partitions definition for a time-partitioned "
+                "asset during a backfill is not supported."
+            )
+
+        else:
+            # Check that all target partitions still exist. If so, the backfill can continue.
+            for target_key in target_partitions_subset.get_partition_keys():
+                if not partitions_def.has_partition_key(
```

Comment:
Will this be O(n^2)? Is there a way to make it O(n)?

Reply:
Assuming that N is the number of partitions on the partitions def. It's not O(N^2) -- each partitions def has its own `has_partition_key` method that is optimized to not require fetching all partitions

Reply:
For `StaticPartitionsDefinition`, my read is that `has_partition_key` looks inside `self._partitions`, which is a sequence and thus O(# partitions) to check if it has a value. Is that off base?

And for dynamic partitions, will this not result in O(target_partitions_subset.get_partition_keys()) calls to the database to get all partition keys?

Reply:
Ah.. yeah, looks like for static partitions defs the call will be O(N^2)

> And for dynamic partitions, will this not result in O(target_partitions_subset.get_partition_keys()) calls to the database to get all partition keys?

And yes, this is true

Reply:
I've amended the behavior to instead use the `AllPartitionsSubset` to build a subset of all valid partitions, then compare against that subset

This should be O(N) now